### PR TITLE
feat: Add histogram metric type

### PR DESCRIPTION
**Pull request by yinggeh**
#### What does the PR do?
Support histogram metric type and its functions.

#### Checklist
- [x] PR title reflects the change and is of format `<commit_type>: <Title>`
- [x] Changes are described in the pull request.
- [x] Related issues are referenced.
- [x] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [x] Added [test plan](#test-plan) and verified test passes.
- [x] Verified that the PR passes existing CI.
- [x] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [x] feat

#### Related PRs:
https://github.com/triton-inference-server/vllm_backend/pull/56
https://github.com/triton-inference-server/core/pull/386
https://github.com/triton-inference-server/server/pull/7525

#### Where should the reviewer start?
n/a

#### Test plan:
n/a

- CI Pipeline ID:
17487728

#### Caveats:
n/a

#### Background
Customer requested histogram metrics from vLLM.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
n/a

### CODE REVIEW

**Comment by oandreeva-nv** [comment]
I think we need general tests for this on python backend as well

File: src/pb_stub.cc [hunk 1]
```diff
@@ -1824,11 +1824,13 @@ PYBIND11_EMBEDDED_MODULE(c_python_backend_utils, module)
   py::class_<Metric, std::shared_ptr<Metric>>(module, "Metric")
       .def("increment", &Metric::SendIncrementRequest)
       .def("set", &Metric::SendSetValueRequest)
+      .def("observe", &Metric::SendObserveRequest)
       .def("value", &Metric::SendGetValueRequest);
 
   py::enum_<MetricKind>(module, "MetricKind")
       .value("COUNTER", MetricKind::kCounter)
       .value("GAUGE", MetricKind::kGauge)
+      .value("HISTOGRAM", MetricKind::kHistogram)
```
Comment (rmccorm4) [inline src/pb_stub.cc:1833]: +1 on @oandreeva-nv comment for testing in python backend too:

Can add to the existing custom metrics tests:
- https://github.com/triton-inference-server/server/blob/main/qa/L0_backend_python/custom_metrics/test.sh
- https://github.com/triton-inference-server/server/blob/main/qa/python_models/custom_metrics/model.py

**Comment by krishung5** [comment]
+1 on the test. Also I think would be great if we can add the documentation for the new type of metrics: https://github.com/triton-inference-server/python_backend/tree/main/examples/custom_metrics

File: src/metric.cc [hunk 2]
```diff
@@ -281,6 +362,13 @@ Metric::SetValue(const double& value)
   THROW_IF_TRITON_ERROR(TRITONSERVER_MetricSet(triton_metric, value));
 }
 
+void
+Metric::Observe(const double& value)
```
Comment (Tabrizian) [inline src/metric.cc:366]: Why not just use `Set`?

Comment (yinggeh) [reply src/metric.cc:366]: In Prometheus, gauge calls `Set` to set itself to a value, while histogram and summary kinds call `Observe` to record a value internally. I think `Set` and `Observe` have different meanings although they are the same API. It also makes the call more clear to people who's familiar with Prometheus metrics. Open to suggestions.
https://github.com/jupp0r/prometheus-cpp/blob/master/core/include/prometheus/histogram.h#L58
https://github.com/jupp0r/prometheus-cpp/blob/master/core/include/prometheus/summary.h#L84

File: src/pb_stub.cc (hunk 1 above)
Comment (yinggeh) [reply src/pb_stub.cc:1833]: Tests added https://github.com/triton-inference-server/server/pull/7525

File: src/metric.cc (hunk 2 above)
Comment (Tabrizian) [reply src/metric.cc:376]: Why we don't do it in core then?

Comment (Tabrizian) [reply src/metric.cc:376]: This is an internal implementation, let's keep it consistent with what we have in core.

Comment (GuanLuo) [reply src/metric.cc:376]: Agree, please keep in sync with the core API.

**Comment by yinggeh** [comment]
> +1 on the test. Also I think would be great if we can add the documentation for the new type of metrics: https://github.com/triton-inference-server/python_backend/tree/main/examples/custom_metrics

Updated. See [commit](https://github.com/triton-inference-server/python_backend/pull/374/commits/e36181229e55ff0b07e12d4cb49949cd55a008b6)

**Review by GuanLuo — APPROVED** [review]

**Review by krishung5 — APPROVED** [review]

**Review by Tabrizian — APPROVED** [review]

File: src/metric.cc (hunk 2 above)
Comment (yinggeh) [reply src/metric.cc:366]: Resolved.

**Pull request merged**